arangod/RocksDBEngine/RocksDBTempStorage.cpp

```diff
@@ -0,0 +1,265 @@
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Julia Puget
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RocksDBTempStorage.h"
+
+#include "Basics/FileUtils.h"
+#include "Basics/VelocyPackHelper.h"
+#include "Basics/application-exit.h"
+#include "Basics/debugging.h"
+#include "Basics/files.h"
+#include "Logger/LogMacros.h"
+#include "RestServer/TemporaryStorageFeature.h"
+#include "RocksDBEngine/RocksDBColumnFamilyManager.h"
+#include "RocksDBEngine/RocksDBSortedRowsStorageContext.h"
+
+#ifdef USE_ENTERPRISE
+#include "Enterprise/RocksDBEngine/EncryptionProvider.h"
+#include "Enterprise/RocksDBEngine/RocksDBEncryptionUtilsEE.h"
+#endif
+
+#include <absl/strings/str_cat.h>
+
+#include <rocksdb/comparator.h>
+#include <rocksdb/db.h>
+#include <rocksdb/env.h>
+#include <rocksdb/env_encryption.h>
+#include <rocksdb/options.h>
+#include <rocksdb/slice_transform.h>
+
+using namespace arangodb;
+
+namespace {
+
+class KeysComparator : public rocksdb::Comparator {
+ public:
+  KeysComparator() = default;
+  int Compare(rocksdb::Slice const& lhs,
+              rocksdb::Slice const& rhs) const override {
+    TRI_ASSERT(lhs.size() >= sizeof(uint64_t));
+    TRI_ASSERT(rhs.size() >= sizeof(uint64_t));
+    // compare first 8 bytes, which is the "context id"
+    int diff = memcmp(lhs.data(), rhs.data(), sizeof(uint64_t));
+    if (diff != 0) {
+      return diff;
+    }
+
+    // move beyond "context id"
+    auto p1 = lhs.data() + sizeof(uint64_t);
+    auto p2 = rhs.data() + sizeof(uint64_t);
+
+    // now compare the running number (used for stable sorts)
+    int diffInId = 0;
+    bool hasPrefixId1 = static_cast<size_t>(p1 - lhs.data()) < lhs.size();
+    bool hasPrefixId2 = static_cast<size_t>(p2 - rhs.data()) < rhs.size();
+    if (hasPrefixId1 && hasPrefixId2) {
+      diffInId = memcmp(p1, p2, sizeof(uint64_t));
+      p1 += sizeof(uint64_t);
+      p2 += sizeof(uint64_t);
+    } else if (hasPrefixId1) {
+      diffInId = 1;
+    } else if (hasPrefixId2) {
+      diffInId = -1;
+    }
+    while (static_cast<size_t>(p1 - lhs.data()) < lhs.size() &&
+           static_cast<size_t>(p2 - rhs.data()) < rhs.size()) {
+      velocypack::Slice slice1(reinterpret_cast<uint8_t const*>(p1));
+      p1 += slice1.byteSize();
+      velocypack::Slice slice2(reinterpret_cast<uint8_t const*>(p2));
+      p2 += slice2.byteSize();
+      char order1 = *p1;
+      char order2 = *p2;
```

Comment:
If either of the two keys has exactly one more VelocyPack item (and not another byte), then we are accessing either lhs or rhs out of bounds here.

Reply:
Maybe this is always guaranteed to be OK, but I do not know this.

Reply:
Added a comment like this:
```
 // here, we expect always pairs of the actual key value to compare (the
    // slice) + the byte that represents the order in which to compare. As we
    // build the arguments ourselves to call this comparator, we can expect that
    // either there would be no more value with order to compare, or the value
    // would be present with the order byte
```

Reply:
ok. Thanks.